File: openstudio-necb/test/test_cli.rb
require_relative 'test_helper'
require_relative '../lib/openstudio_necb/cli'
require 'stringio'
require 'tmpdir'

# The CLI is exercised IN-PROCESS (CLI.run returns an Integer and never exits),
# so these are real tests rather than subprocess smoke checks. Only the last one
# simulates; everything else is a parse/pre-flight assertion and runs in seconds.
class TestCLI < Minitest::Test
  include FixtureHelper

  def setup
    @out = StringIO.new
    @err = StringIO.new
  end

  def run_cli(*argv)
    OpenStudioNECB::CLI.run(argv, out: @out, err: @err)
  end

  def weather_args
    ['--epw', EPW, '--ddy', DDY]
  end

  # --- usage -------------------------------------------------------------

  def test_help_exits_zero_and_documents_the_required_arguments
    assert_equal(0, run_cli('--help'))
    assert_match(/Usage: necb-compliance MODEL\.osm --epw FILE/, @out.string)
    assert_match(/--space-type/, @out.string)
  end

  def test_no_model_is_a_usage_error
    assert_equal(2, run_cli)
    assert_match(/no model given/, @err.string)
  end

  def test_missing_model_file_is_a_usage_error
    assert_equal(2, run_cli('/nonexistent/model.osm', '--epw', EPW))
    assert_match(%r{model not found: /nonexistent/model\.osm}, @err.string)
  end

  def test_missing_epw_is_a_usage_error
    assert_equal(2, run_cli(FIXTURE))
    assert_match(/no --epw given/, @err.string)
  end

  # attach_weather_and_hdd! would also raise for this, but only after the model
  # load — a fat-fingered path must not cost the user that wait.
  def test_missing_ddy_sibling_is_caught_before_the_model_loads
    Dir.mktmpdir do |dir|
      lonely = File.join(dir, 'weather.epw')
      FileUtils.cp(EPW, lonely)
      assert_equal(2, run_cli(FIXTURE, '--epw', lonely))
      assert_match(/ddy not found/, @err.string)
      assert_match(/pass --ddy explicitly/, @err.string)
    end
  end

  def test_unknown_flag_is_a_usage_error
    assert_equal(2, run_cli(FIXTURE, '--epw', EPW, '--nope'))
    assert_match(/invalid option/, @err.string)
  end

  def test_bad_vintage_is_rejected_by_the_parser
    assert_equal(2, run_cli(FIXTURE, *weather_args, '--vintage', '2011'))
  end

  # --- the pre-flight ----------------------------------------------------

  # The shared fixture is ASHRAE-tagged with no standardsSpaceType, so this is a
  # fast, simulation-free assertion that the refusal is reported as its OWN exit
  # code rather than collapsed into a generic error.
  def test_untagged_model_is_refused_with_exit_3_and_actionable_advice
    Dir.mktmpdir do |dir|
      code = run_cli(FIXTURE, *weather_args, '--simulate', 'none', '--hdd', '3890',
                     '--storeys', '1', '-o', File.join(dir, 'run'))
      assert_equal(3, code)
      assert_match(/REJECTED before any simulation ran/, @err.string)
      assert_match(/pre-flight FAILED/, @err.string)
      assert_match(/--space-type/, @err.string, 'must name the on-ramp that fixes it')
    end
  end

  def test_preflight_error_is_an_argument_error_so_existing_rescues_still_work
    assert_operator(OpenStudioNECB::PreflightError, :<, ArgumentError)
  end

  # --- no determination --------------------------------------------------

  def test_simulate_none_reports_no_determination_not_a_verdict
    Dir.mktmpdir do |dir|
      code = run_cli(FIXTURE, *weather_args, '--simulate', 'none', '--hdd', '3890',
                     '--storeys', '1', '--no-report', '--quiet',
                     '--space-type', 'Space Function/Office enclosed > 25 m2',
                     '-o', File.join(dir, 'run'))
      assert_equal(6, code)
      assert_match(/NO DETERMINATION/, @out.string)
      refute_match(/VERDICT: COMPLIANT/, @out.string)
    end
  end

  # --- end to end --------------------------------------------------------

  # The one test that simulates. A shortened run period must NEVER be reported
  # as compliant: evaluate() still returns a boolean and only sets
  # report['annual'] = false, so the CLI has to override it.
  def test_quick_annual_run_produces_a_report_but_refuses_to_call_it_a_determination
    skip('openstudio CLI not available') unless OpenStudioNECB::Runner.openstudio_cli?

    Dir.mktmpdir do |dir|
      run_dir = File.join(dir, 'run')
      code = run_cli(FIXTURE, *weather_args, '--simulate', 'annual', '--quick',
                     '--hdd', '3890', '--storeys', '1', '--quiet',
                     '--space-type', 'Space Function/Office enclosed > 25 m2',
                     '--shw-fuel', 'NaturalGas', '--hvac-system', 'Baseboard gas boiler',
                     '-o', run_dir)

      assert_equal(6, code, 'a shortened run is NOT a code determination')
      assert_match(/NOT A CODE-COMPLIANT DETERMINATION/, @out.string)
      refute_match(/VERDICT: COMPLIANT/, @out.string)

      assert(File.size?(File.join(run_dir, 'compliance_report.html')), 'HTML report must be written')
      report = JSON.parse(File.read(File.join(run_dir, 'report.json'), encoding: 'UTF-8'))
      assert_equal(false, report['annual'])
      assert_operator(report.dig('proposed', 'total_site_kwh'), :>, 0)
      assert_operator(report.dig('reference', 'total_site_kwh'), :>, 0)
      assert_operator(report.dig('proposed', 'eui_kwh_per_m2'), :>, 0)
      assert_operator(report.dig('reference', 'eui_kwh_per_m2'), :>, 0)
    end
  end

  # --- output ------------------------------------------------------------

  # The Windows console is CP437/CP1252; a superscript or an em dash on stdout
  # would be the tool breaking rather than the tool reporting.
  def test_stdout_is_pure_ascii
    Dir.mktmpdir do |dir|
      run_cli(FIXTURE, *weather_args, '--simulate', 'none', '--hdd', '3890',
              '--storeys', '1', '--no-report', '--quiet',
              '--space-type', 'Space Function/Office enclosed > 25 m2',
              '-o', File.join(dir, 'run'))
      assert(@out.string.ascii_only?, "stdout carried non-ASCII: #{@out.string.scan(/[^\x00-\x7F]/).uniq.inspect}")
    end
  end

  def test_help_output_is_pure_ascii
    run_cli('--help')
    assert(@out.string.ascii_only?)
  end
  # --- packaging invariants ----------------------------------------------

  # The Windows installer ships everything EXCEPT the two priced RS-Means-derived
  # tables; costing still works when the user points --costs-csv at their own.
  # This pins that the COMPLIANCE path never touches them, so a future eager read
  # breaks CI here rather than shipping an installer that dies at run time.
  def test_compliance_runs_without_the_priced_costing_tables
    # Explicit gem path, not derived from HVAC_FIXTURES — the fixtures moved
    # to btap-modeling while the priced tables stay with hvac costing until
    # btap-costing exists.
    costing = File.expand_path('../../openstudio-hvac/lib/openstudio_hvac/data/costing', __dir__)
    priced = Dir.glob(File.join(costing, '{costs,costs_local_factors}.csv'))
    # NOT a skip: this checkout vendors both, and a silently-skipping packaging
    # gate is exactly the green-but-vacuous failure this repo already documents.
    assert_equal(2, priced.size, "expected both priced tables in #{costing}")

    moved = priced.map { |f| [f, "#{f}.hidden"] }
    begin
      moved.each { |src, dst| FileUtils.mv(src, dst) }
      Dir.mktmpdir do |dir|
        code = run_cli(FIXTURE, *weather_args, '--simulate', 'none', '--hdd', '3890',
                       '--storeys', '1', '--no-report', '--quiet',
                       '--space-type', 'Space Function/Office enclosed > 25 m2',
                       '-o', File.join(dir, 'run'))
        assert_equal(6, code, "the compliance path must not need the priced tables: #{@err.string}")
      end
    ensure
      moved.each { |src, dst| FileUtils.mv(dst, src) if File.exist?(dst) }
    end
  end

  # --- weather -----------------------------------------------------------

  def test_list_cities_needs_no_model
    assert_equal(0, run_cli('--list-cities'))
    assert_match(/toronto/, @out.string)
  end

  def test_unknown_city_lists_what_is_available
    assert_equal(2, run_cli(FIXTURE, '--city', 'atlantis'))
    assert_match(/unknown city: atlantis/, @err.string)
    assert_match(/known: /, @err.string)
  end

  def test_city_and_epw_are_mutually_exclusive
    assert_equal(2, run_cli(FIXTURE, '--city', 'toronto', '--epw', EPW))
    assert_match(/mutually exclusive/, @err.string)
  end

  # --city must supply the DDY too — attach_weather! hard-requires it, and the
  # .stat beside it is what Climate.hdd18 reads before falling back to Table C-1.
  def test_city_resolves_to_a_file_whose_ddy_and_stat_sit_beside_it
    epw = OpenStudioNECB::CLI::Weather.available['toronto']
    refute_nil(epw, 'the fixture weather should be discoverable as "toronto"')
    assert(File.exist?(epw.sub(/\.epw\z/, '.ddy')), 'no .ddy beside the resolved EPW')
    assert(File.exist?(epw.sub(/\.epw\z/, '.stat')), 'no .stat beside the resolved EPW')
  end
  # The bug this pins: --list-cities reported "No weather files found in this
  # installation" on a real Windows install while every test here passed,
  # because in a SOURCE CHECKOUT the openstudio-hvac fixtures path answered and
  # masked the wrong depth. The packaged tree puts the gems one level deeper
  # (under gems/), so the relative path needs 4 ups, not 3.
  #
  # Layout-specific paths need a test per LAYOUT, not per platform — this builds
  # the packaged layout in a tmpdir rather than trusting the checkout.
  def test_weather_is_discoverable_in_the_PACKAGED_layout
    Dir.mktmpdir do |root|
      gem_lib = File.join(root, 'gems/openstudio-necb/lib/openstudio_necb')
      FileUtils.mkdir_p([gem_lib, File.join(root, 'weather')])
      %w[epw ddy stat].each do |ext|
        FileUtils.cp(File.join(File.dirname(EPW), "#{File.basename(EPW, '.epw')}.#{ext}"),
                     File.join(root, 'weather'))
      end

      found = OpenStudioNECB::CLI::Weather.search
                                          .map { |d| File.expand_path(d.to_s) }
      packaged = File.expand_path('../../../../weather', gem_lib)
      assert_equal(File.join(root, 'weather'), packaged, 'sanity: 4 ups is the packaged root')
      assert_includes(found, File.expand_path('../../../../weather', __dir__.sub('/test', '/lib/openstudio_necb')),
                      'the 4-up packaged path must be searched')
    end
  end

  # NECB_HOME is what the launcher actually sets, so it must win over any
  # checkout path that happens to exist on the same machine.
  def test_necb_home_takes_precedence
    Dir.mktmpdir do |home|
      FileUtils.mkdir_p(File.join(home, 'weather'))
      FileUtils.cp(EPW, File.join(home, 'weather', 'CAN_XX_Somewhere.123456_CWEC2020.epw'))
      begin
        ENV['NECB_HOME'] = home
        assert_equal(File.join(home, 'weather'), OpenStudioNECB::CLI::Weather.search.first)
        assert_equal(File.join(home, 'weather', 'CAN_XX_Somewhere.123456_CWEC2020.epw'),
                     OpenStudioNECB::CLI::Weather.available['somewhere'])
      ensure
        ENV.delete('NECB_HOME')
      end
    end
  end
end

File: python/tests/necb/test_cli.py
"""Port of btap-necb/test/test_cli.rb: the CLI exercised IN-PROCESS (cli.run
returns an int and never exits), so these are real tests rather than
subprocess smoke checks. Only one test simulates; everything else is a
parse/pre-flight assertion and runs in seconds.

Parser-specific stderr wording differs from Ruby's optparse (argparse says
"invalid option: --nope" via our own funnel) — the exit codes and the
semantics are what is pinned."""

import io
import json
import os
import shutil
import subprocess
import sys
import tempfile
import unittest
from pathlib import Path

from btap.necb import cli
from tests.necb.support import DDY, EPW, FIXTURE_OSM, needs_engine, needs_sdk

FIXTURE = str(FIXTURE_OSM)


class CLICase(unittest.TestCase):
    def setUp(self):
        self.out = io.StringIO()
        self.err = io.StringIO()

    def run_cli(self, *argv):
        return cli.run(list(argv), out=self.out, err=self.err)

    def weather_args(self):
        return ["--epw", str(EPW), "--ddy", str(DDY)]


class TestCLIUsage(CLICase):
    def test_help_exits_zero_and_documents_the_required_arguments(self):
        self.assertEqual(0, self.run_cli("--help"))
        self.assertRegex(self.out.getvalue(),
                         r"btap-compliance MODEL\.osm --epw FILE")
        self.assertRegex(self.out.getvalue(), r"--space-type")

    def test_no_model_is_a_usage_error(self):
        self.assertEqual(2, self.run_cli())
        self.assertRegex(self.err.getvalue(), r"no model given")

    def test_missing_model_file_is_a_usage_error(self):
        self.assertEqual(2, self.run_cli("/nonexistent/model.osm",
                                         "--epw", str(EPW)))
        self.assertRegex(self.err.getvalue(),
                         r"model not found: /nonexistent/model\.osm")

    def test_missing_epw_is_a_usage_error(self):
        self.assertEqual(2, self.run_cli(FIXTURE))
        self.assertRegex(self.err.getvalue(), r"no --epw given")

    def test_missing_ddy_sibling_is_caught_before_the_model_loads(self):
        # The pipeline's weather guard would also raise for this, but only
        # after the model load — a fat-fingered path must not cost that wait.
        with tempfile.TemporaryDirectory() as dir:
            lonely = os.path.join(dir, "weather.epw")
            shutil.copy(str(EPW), lonely)
            self.assertEqual(2, self.run_cli(FIXTURE, "--epw", lonely))
            self.assertRegex(self.err.getvalue(), r"ddy not found")
            self.assertRegex(self.err.getvalue(), r"pass --ddy explicitly")

    def test_unknown_flag_is_a_usage_error(self):
        self.assertEqual(2, self.run_cli(FIXTURE, "--epw", str(EPW), "--nope"))
        self.assertRegex(self.err.getvalue(), r"invalid option")

    def test_bad_vintage_is_rejected_by_the_parser(self):
        self.assertEqual(2, self.run_cli(FIXTURE, *self.weather_args(),
                                         "--vintage", "2011"))


@needs_sdk
class TestCLIPreflight(CLICase):
    def test_untagged_model_is_refused_with_exit_3_and_actionable_advice(self):
        # The shared fixture is ASHRAE-tagged with no standardsSpaceType, so
        # this is a fast, simulation-free assertion that the refusal is
        # reported as its OWN exit code rather than collapsed into a generic
        # error.
        with tempfile.TemporaryDirectory() as dir:
            code = self.run_cli(FIXTURE, *self.weather_args(),
                                "--simulate", "none", "--hdd", "3890",
                                "--storeys", "1", "-o",
                                os.path.join(dir, "run"))
            self.assertEqual(3, code)
            self.assertRegex(self.err.getvalue(),
                             r"REJECTED before any simulation ran")
            self.assertRegex(self.err.getvalue(), r"pre-flight FAILED")
            self.assertRegex(self.err.getvalue(), r"--space-type",
                             "must name the on-ramp that fixes it")

    def test_preflight_error_is_a_value_error_so_existing_handlers_work(self):
        from btap.necb.compliance import PreflightError

        self.assertTrue(issubclass(PreflightError, ValueError))

    def test_simulate_none_reports_no_determination_not_a_verdict(self):
        with tempfile.TemporaryDirectory() as dir:
            code = self.run_cli(
                FIXTURE, *self.weather_args(), "--simulate", "none",
                "--hdd", "3890", "--storeys", "1", "--no-report", "--quiet",
                "--space-type", "Space Function/Office enclosed > 25 m2",
                "-o", os.path.join(dir, "run"))
            self.assertEqual(6, code)
            self.assertRegex(self.out.getvalue(), r"NO DETERMINATION")
            self.assertNotRegex(self.out.getvalue(), r"VERDICT: COMPLIANT")

    def test_stdout_is_pure_ascii(self):
        # The Windows console is CP437/CP1252; a superscript or an em dash on
        # stdout would be the tool breaking rather than the tool reporting.
        with tempfile.TemporaryDirectory() as dir:
            self.run_cli(
                FIXTURE, *self.weather_args(), "--simulate", "none",
                "--hdd", "3890", "--storeys", "1", "--no-report", "--quiet",
                "--space-type", "Space Function/Office enclosed > 25 m2",
                "-o", os.path.join(dir, "run"))
            text = self.out.getvalue()
            bad = sorted({c for c in text if not c.isascii()})
            self.assertTrue(text.isascii(), f"stdout carried non-ASCII: {bad}")

    def test_compliance_runs_without_the_priced_costing_tables(self):
        # The installer ships everything EXCEPT the two priced
        # RS-Means-derived tables; costing still works when the user points
        # --costs-csv at their own. This pins that the COMPLIANCE path never
        # touches them, so a future eager read breaks CI here rather than
        # shipping a distribution that dies at run time.
        #
        # PROBE MECHANICS (learned the hard way): Ruby's version renamed the
        # shared CSVs on disk and restored them after — safe there because
        # its runner forks per GEM, so nothing else could read them in the
        # window. Under pytest-xdist the costing suites run CONCURRENTLY in
        # sibling workers, and hiding the real files crashed five of them in
        # CI. The probe is now PROCESS-LOCAL and stronger: a subprocess runs
        # the CLI with builtins.open guarded so ANY read of either priced
        # filename raises — nothing on disk moves, and eager reads through
        # any resolution layer are caught, not just the vendored copies.
        costing = Path(cli.__file__).parents[1] / "costing" / "data"
        # NOT a skip: this checkout vendors both, and a silently-skipping
        # packaging gate is exactly the green-but-vacuous failure this repo
        # already documents.
        for name in ("costs.csv", "costs_local_factors.csv"):
            self.assertTrue((costing / name).exists(),
                            f"expected the priced table {name} in {costing}")

        wrapper = (
            "import builtins, os.path, sys\n"
            "_real = builtins.open\n"
            "def guarded(file, *a, **k):\n"
            "    if os.path.basename(str(file)) in ('costs.csv',"
            " 'costs_local_factors.csv'):\n"
            "        raise AssertionError('the compliance path read a priced"
            " table: %s' % file)\n"
            "    return _real(file, *a, **k)\n"
            "builtins.open = guarded\n"
            "from btap.necb import cli\n"
            "sys.exit(cli.run(sys.argv[1:]))\n")
        with tempfile.TemporaryDirectory() as dir:
            script = os.path.join(dir, "guarded_cli.py")
            Path(script).write_text(wrapper, encoding="utf-8")
            env = dict(os.environ)
            python_root = str(Path(cli.__file__).parents[2])
            env["PYTHONPATH"] = os.pathsep.join(
                p for p in (python_root, env.get("PYTHONPATH")) if p)
            proc = subprocess.run(
                [sys.executable, script,
                 FIXTURE, *self.weather_args(), "--simulate", "none",
                 "--hdd", "3890", "--storeys", "1", "--no-report",
                 "--quiet", "--space-type",
                 "Space Function/Office enclosed > 25 m2",
                 "-o", os.path.join(dir, "run")],
                capture_output=True, text=True, env=env, timeout=600)
            self.assertEqual(6, proc.returncode,
                             "the compliance path must not need the priced "
                             f"tables: {proc.stderr[-2000:]}")
            self.assertNotIn("priced table", proc.stderr,
                             "a priced table WAS read")


class TestCLIOutput(CLICase):
    def test_help_output_is_pure_ascii(self):
        self.run_cli("--help")
        self.assertTrue(self.out.getvalue().isascii())


class TestCLIWeather(CLICase):
    def test_list_cities_needs_no_model(self):
        self.assertEqual(0, self.run_cli("--list-cities"))
        self.assertRegex(self.out.getvalue(), r"toronto")

    def test_unknown_city_lists_what_is_available(self):
        self.assertEqual(2, self.run_cli(FIXTURE, "--city", "atlantis"))
        self.assertRegex(self.err.getvalue(), r"unknown city: atlantis")
        self.assertRegex(self.err.getvalue(), r"known: ")

    def test_city_and_epw_are_mutually_exclusive(self):
        self.assertEqual(2, self.run_cli(FIXTURE, "--city", "toronto",
                                         "--epw", str(EPW)))
        self.assertRegex(self.err.getvalue(), r"mutually exclusive")

    def test_city_resolves_to_a_file_whose_ddy_and_stat_sit_beside_it(self):
        # --city must supply the DDY too — attach_weather hard-requires it,
        # and the .stat beside it is what the climate resolver reads before
        # falling back to Table C-1.
        epw = cli.Weather.available().get("toronto")
        self.assertIsNotNone(
            epw, 'the fixture weather should be discoverable as "toronto"')
        self.assertTrue(os.path.exists(epw[:-4] + ".ddy"),
                        "no .ddy beside the resolved EPW")
        self.assertTrue(os.path.exists(epw[:-4] + ".stat"),
                        "no .stat beside the resolved EPW")

    def test_btap_home_takes_precedence(self):
        # BTAP_HOME is what a launcher actually sets, so it must win over any
        # checkout path that happens to exist on the same machine.
        with tempfile.TemporaryDirectory() as home:
            os.makedirs(os.path.join(home, "weather"))
            shutil.copy(str(EPW), os.path.join(
                home, "weather", "CAN_XX_Somewhere.123456_CWEC2020.epw"))
            os.environ["BTAP_HOME"] = home
            try:
                self.assertEqual(os.path.join(home, "weather"),
                                 cli.Weather.search()[0])
                self.assertEqual(
                    os.path.join(home, "weather",
                                 "CAN_XX_Somewhere.123456_CWEC2020.epw"),
                    cli.Weather.available()["somewhere"])
            finally:
                del os.environ["BTAP_HOME"]


@needs_engine
class TestCLIEndToEnd(CLICase):
    def test_quick_annual_run_refuses_to_call_itself_a_determination(self):
        # The one test that simulates. A shortened run period must NEVER be
        # reported as compliant: the pipeline's evaluate still returns a
        # boolean and only sets report['annual'] = False, so the CLI has to
        # override it.
        with tempfile.TemporaryDirectory() as dir:
            run_dir = os.path.join(dir, "run")
            code = self.run_cli(
                FIXTURE, *self.weather_args(), "--simulate", "annual",
                "--quick", "--hdd", "3890", "--storeys", "1", "--quiet",
                "--space-type", "Space Function/Office enclosed > 25 m2",
                "--shw-fuel", "NaturalGas",
                "--hvac-system", "Baseboard gas boiler",
                "-o", run_dir)

            self.assertEqual(6, code,
                             "a shortened run is NOT a code determination: "
                             + self.err.getvalue()[-2000:])
            self.assertRegex(self.out.getvalue(),
                             r"NOT A CODE-COMPLIANT DETERMINATION")
            self.assertNotRegex(self.out.getvalue(), r"VERDICT: COMPLIANT")

            html_path = os.path.join(run_dir, "compliance_report.html")
            self.assertTrue(os.path.getsize(html_path),
                            "HTML report must be written")
            report = json.loads(Path(run_dir, "report.json").read_text(
                encoding="utf-8"))
            self.assertEqual(False, report["annual"])
            self.assertGreater(report["proposed"]["total_site_kwh"], 0)
            self.assertGreater(report["reference"]["total_site_kwh"], 0)
            self.assertGreater(report["proposed"]["eui_kwh_per_m2"], 0)
            self.assertGreater(report["reference"]["eui_kwh_per_m2"], 0)


if __name__ == "__main__":
    unittest.main()
